File: TP/TP-C/TP2/pub.c
//  Weather update server
//  Binds PUB socket to tcp://*:5556
//  Publishes random weather updates

#include "zhelpers.h"

int main (int arg, char *argv [])
{
    //  Prepare our context and publisher
    void *context = zmq_ctx_new ();
    void *publisher = zmq_socket (context, ZMQ_PUB);
    char bind[15] = "tcp://*:";
    strcat(bind, argv[1]);
    int rc = zmq_bind(publisher, bind);
    assert (rc == 0);

    //  Initialize random number generator
    srandom ((unsigned) time (NULL));
    while (1) {
        //  Get values that will fool the boss
        int zipcode, temperature, relhumidity;
        zipcode     = randof (100000);
        temperature = randof (215) - 80;
        relhumidity = randof (50) + 10;

        //  Send message to all subscribers
        char update [20];
        sprintf (update, "%d %d %d", zipcode, temperature, relhumidity);
        s_send (publisher, update);
    }
    zmq_close (publisher);
    zmq_ctx_destroy (context);
    return 0;
}

File: TP/TP-Python/TP2/pub.py


#
#   Weather update server
#   Binds PUB socket to tcp://*:5556
#   Publishes random weather updates
#

import zmq
from random import randrange
import sys

# Check for port argument
if len(sys.argv) != 2:
    print("Usage: python pub.py <port>")
    sys.exit(1)

port = sys.argv[1]
bind_address = f"tcp://*:{port}"

context = zmq.Context()
socket = context.socket(zmq.PUB)
socket.bind(bind_address)

while True:
    zipcode = randrange(1, 100000)
    temperature = randrange(-80, 135)
    relhumidity = randrange(10, 60)

    socket.send_string(f"{zipcode} {temperature} {relhumidity}")

